<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [75]:
connection = psycopg2.connect('[ДАННЫЕ УДАЛЕНЫ]')

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
query_3_1 = f'''SELECT COUNT(*) 
           FROM public.vacancies
        '''
        
pd.read_sql_query(query_3_1, connection)

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1620135908.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query, connection)


count
0  49197

Всего в базе данных содержится информация о 49 197 вакансиях. В последующих расчётах я буду округлять это значение до тысяч (49 000 вакансий).

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
query_3_2 = f'''SELECT COUNT(*) 
           FROM public.employers
        '''
        
pd.read_sql_query(query_3_2, connection)

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\275374076.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query, connection)


count
0  23501

В базе данных содержится информация о вакансиях, принадлежащих 23 501 работодателю. В последующих расчётах я также буду округлять это значение до тысяч (24 000 работодателей).

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [7]:
query_3_3 = f'''SELECT COUNT(*) 
           FROM public.areas
        '''
        
pd.read_sql_query(query_3_3, connection)

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\87212700.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query, connection)


count
0   1362

В базе данных содержится информация о вакансиях из 1362 регионов. В дальнейшем я буду округлять это значение до сотен (1400 регионов).

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [9]:
query_3_4 = f'''SELECT COUNT(*) 
           FROM public.industries
        '''
        
pd.read_sql_query(query_3_4, connection)

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1260782755.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query, connection)


count
0    294

В базе данных содержится информация о вакансиях, относящихся к 294 сферам деятельности. В дальнейшем я буду округлять это значение до сотен (300 сфер деятельности).

***

## Вывод по Юниту 3

Итак, в исследуемой базе данных содержатся сведения о 49 тысячах индивидуальных вакансий, принадлежащих 23 тысячам работодателей,
и относящихся к почти трём сотням сфер деятельности. Если при создании данной выборки вакансий не было допущено каких-то намеренных
искажений, она должна быть достаточно репрезентативна, чтобы на её основании делать адекватные реальности суждения
о ситуации на интернет-рынке рабочих мест.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько () вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [68]:
query_4_1 = '''SELECT COUNT(v.id), a.name
            FROM public.vacancies v
            JOIN public.areas a ON v.area_id = a.id
            GROUP BY a.name
            ORDER BY 1 DESC
            LIMIT 10
            '''
            
result = pd.read_sql_query(query_4_1, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\148310811.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_1, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы
5   1698     Екатеринбург
6   1670  Нижний Новгород
7   1415           Казань
8   1301        Краснодар
9   1144           Самара

Разумеется, работать с количеством вакансий в каждом из всех четырнадцати сотен регионов, представленных таблице, будет практически невозможно.
Поэтому я представил в вывде только первые десять (по убыванию количества вакансий) регионов. В данной базе данных это: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы, Екатеринбург, Нижний Новгород, Казань, Краснодар, Самара. Тремя самыми "богатыми" на вакансии среди них являются Москва (5333 вакансии), Санкт-Петербург (2851 вакансия) и Минск (2112 вакансий);замыкают же этот топ-десять Казань (1415 вакансий), Краснодар (1301 вакансия) и Самара (1144 вакансии).

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [13]:
query_4_2 = '''SELECT COUNT(*)
            FROM public.vacancies v
            WHERE v.salary_from IS NOT NULL 
                  OR v.salary_to IS NOT NULL
            '''
            
result = pd.read_sql_query(query_4_2, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\2741981258.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_2, connection)


count
0  24073

Хотя бы одно из двух полей с зарплатой заполнено у 24 073 вакансий, что можно округлить до 24 000 вакансий.

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [69]:
query_4_3 = '''SELECT 
                    ROUND(AVG(v.salary_from)),
                    ROUND(AVG(v.salary_to))
            FROM public.vacancies v
            '''
            
result = pd.read_sql_query(query_4_3, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1039421947.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_3, connection)


round     round
0  71065.0  110537.0

Среднее значение нижней границы заработной вилки в базе данных составляет 71 065 рублей.
Среднее значение ;же верхней границы - 110 537 рублей.

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [17]:
query_4_4 = '''SELECT 
                    COUNT(*),
                    v.schedule,
                    v.employment
            FROM public.vacancies v
            GROUP BY 
                     v.schedule,
                     v.employment
            ORDER BY 1 DESC
            LIMIT 5
            '''
            
result = pd.read_sql_query(query_4_4, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\128273891.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_4, connection)


count          schedule           employment
0  35367       Полный день     Полная занятость
1   7802  Удаленная работа     Полная занятость
2   1593     Гибкий график     Полная занятость
3   1312  Удаленная работа  Частичная занятость
4    940    Сменный график     Полная занятость

Комбинации типов рабочего графика и трудоустройства и соответствующее им количество вакансий в базе данных: Полный день/Полная занятость (35 367 вакансий), Удалённая работа/Полная занятость (7802 вакансии), Гибкий график/Полная занятость (1593 вакансии), Удалённая работа/Частичная занятость (1312 вакансий) и Сменный график/Полная занятость (940 вакансий).

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [19]:
query_4_5 = '''SELECT 
                    COUNT(*),
                    v.experience
            FROM public.vacancies v
            GROUP BY v.experience
            ORDER BY 1 ASC
            '''
            
result = pd.read_sql_query(query_4_5, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\452533958.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_5, connection)


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

Категории опыта соискателей, представленные в вакансиях, по возрастанию: Самая редкая - "Более 6 лет" (1337 вакансий), затем категория "Нет опыта" (7197 вакансий), затем "От 3 до 6 лет" (14 511 вакансий), и наконец, наиболее представленная в базе данных категория - "От 1 года до 3 лет" (26 152 вакансии).

***

## Выводы по Юниту 4

Регион-"чемпион" по количеству вакансий, лидирующий с двукратным отрывом по сравнению со следующим за ним регионом - это Москва.
Ничего аномального в этом нет - это естественное следствие статуса Москвы как столицы, крупнейшего города и центра экономической
жизни России. Можно сделать вывод, что выборка вакансий, содержавшаяся в исходной БД, адекватно отражает экономическую и демографическую структуру РФ, а значит, является репрезентативной для анализа каких-либо закономерностей на интернет-рынке труда страны.

Примечательным является отсутствие указанной заработной платы у больше, чем половины всех вакансий из имеющихся 
(хотя бы одна граница диапазона возможной заработной платы указана у 24 тысяч вакансий при том, что всего их в базе данных содержится 49 тысяч).
Можно предположить, что в данный момент времени работодатели предпочитали договариваться с новыми специалистами о заработной плате в индивидуальном порядке на самом собеседовании. Скорее всего, в момент заполнения базы данных общий уровень зарплат специалистов на рынке труда вырос по какой-то причине, но большинство работодателей было не в состоянии повысить зарплату уже работавшему у них персоналу, поэтому им пришлось предлагать повышенную зарплату новым специалистам, но не афишировать это публично, чтобы не демотивировать тех, кто уже работал на них. Посмотрим на среднюю предлагаемую зарплату в тех вакансиях, для которых она указана, чтобы проверить это предположение.

Средние нижняя и верхняя границы указываемой работодателями заработной платы в вакнсиях в момент заполнения базы данных составляли 71 тысячу и 
110.5 тысяч рублей соответственно. Следовательно, средняя предлагаемая заработная плата на момент заполнения базы данных составляла 
(71 + 110.5) / 2 = примерно 90.8 тысяч рублей. В настоящий момент (апрель 2023) средняя предлагаемая заработная плата на сайте hh.ru составляет 
58.3 тысячи рублей, что почти на треть меньше, чем аналогичный показатель на момент заполнения базы данных. Учитывая большое количество вакансий без указания заработной платы (см. абзац выше) разумно будет предположить, что средняя предлагаемая зарплата на настоящий момент примерно отражает норму, а средний уровень предлагаемой в вакансиях зарплаты на момент заполнения базы данных являлся аномально повышенным по какой-то причине.

Абсолютное большинство вакансий - почти три четверти из представленных в базе - требуют полной занятости в течение полного дня, что неудивительно,
так как всё ещё является стандартным и распространённым способом занятости. Стоит, однако, отметить, что стремительный рост популярности в настоящий момент удалённой работы нашёл своё отражение в данных - около 19 процентов всех вакансий в базе данных рассчитаны на "удалёнку", при этом в среднем одна из девяти вакансий, рассчитанных на удалённую работу,также предполагает частичную занятость. А вот вакансии, предполагаяющие гибкий или сменный график, популярностью у работодателей не пользуются - обе эти категории вместе составляют всего несколько процентов от общего объёма выборки.

Большинство (чуть больше половины, 53%) вакансий в базе расчитаны на соискателей с опытом работы от одного года до трёх лет. Вторая по популярности у работодателей категория опыта соискателей - от трёх до шести лет (на неё рассчитаны чуть меньше трети от всех вакансий в базе). Интересно, что вакансий, рассчитанных на самых опытных работников (с опытом работы более шести лет) в выборке меньше всего - всего лишь несколько процентов от общего числа вакансий. Скорее всего, это связано с тем, что настолько опытных специалистов большинству работодателей обычно требуется очень небольшое количество. На людей без опыта работы рассчитано около одной седьмой от всех вакансий. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [21]:
query_5_1 = '''(SELECT 
                    COUNT(*),
                    e.name
            FROM public.vacancies v
            JOIN public.employers e ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY 1 DESC
            LIMIT 1)
            UNION ALL
            (SELECT 
                    COUNT(*),
                    e.name
            FROM public.vacancies v
            JOIN public.employers e ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY 1 DESC
            OFFSET 4
            LIMIT 1 )           
            '''
            
result = pd.read_sql_query(query_5_1, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\645678260.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_6, connection)


count           name
0   1933         Яндекс
1    331  Газпром нефть

На первом месте по количеству открытых вакансий находится "Яндекс". На пятом месте - "Газпром нефть".

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [23]:
query_5_2 = '''WITH areas_vacancies AS
             (SELECT a.name
             FROM 
                  public.areas a
                  LEFT JOIN public.vacancies v ON a.id = v.area_id
             GROUP BY a.name
             HAVING COUNT(DISTINCT v.id) = 0)
             SELECT 
                    a.name,
                    COUNT(e.id)
             FROM
                    public.areas a
                    JOIN areas_vacancies temp ON temp.name = a.name
                    JOIN public.employers e ON a.id = e.area
             GROUP BY a.name
             ORDER BY COUNT(e.id) DESC
             LIMIT 1
            '''
            
result = pd.read_sql_query(query_5_2, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\2302336104.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_7, connection)


name  count
0  Россия    410

Как можно увидеть, при создании структуры базы данных была допущена ошибка - в одной и той же таблице регионов содержится информация как о городах, так и о странах. Так как идентификатор региона в таблице вакансий, очевидно, был сделан на строго уровне городов, всем строкам в таблице регионов, в которых были указаны страны, не соответствует ни одной вакансии. Однако в таблице работодателей, по видимому, указывать страну в качестве региона можно. Из-за этого "город", в котором нет вакансий, но есть больше всего работодателей среди подобных "городов" - это Россия.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [71]:
query_5_3 = '''SELECT 
                    e.name,
                    COUNT(DISTINCT v.area_id) area_count
             FROM
                    public.employers e
                    JOIN public.vacancies v ON e.id = v.employer_id 
             GROUP BY e.name
             ORDER BY COUNT(e.name) DESC
             LIMIT 10
            '''
            
result = pd.read_sql_query(query_5_3, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\3776117416.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_8, connection)


name  area_count
0                 Яндекс         181
1             Ростелеком         152
2               Тинькофф          43
3                   СБЕР          24
4          Газпром нефть          24
5            ИК СИБИНТЕК          46
6                    МТС          55
7                DataArt           2
8  Совкомбанк Технологии          26
9             Первый Бит          43

Топ-10 работодателей по количеству регионов, в которых они открывают свои вакансии, по убыванию: "Яндекс", "Ростелеком", "Тинькофф", "СБЕР", "Газпром нефть", "ИК СИБИНТЕК", "МТС", "DataArt", "Совкомбанк Технологии", "Первый Бит".

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [27]:
query_5_4 = '''SELECT COUNT(e.id)
             FROM 
                  public.employers e
                  LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
             WHERE ei.employer_id IS NULL     
            '''
            
result = pd.read_sql_query(query_5_4, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\3089506585.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_9, connection)


count
0   8419

В базе данных сфера деятельности не указана у 8419 работодателей.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [29]:
query_5_5 = '''SELECT e.name
             FROM 
                  public.employers e
                  JOIN public.employers_industries ei ON e.id = ei.employer_id
             GROUP BY e.id
             HAVING COUNT(ei.industry_id) = 4
             ORDER BY e.name
             OFFSET 2
             LIMIT 1    
            '''
            
result = pd.read_sql_query(query_5_5, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\613090776.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_10, connection)


name
0  2ГИС

Название этого работодателя - 2ГИС.

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [31]:
query_5_6 = '''SELECT COUNT(e.id)
             FROM 
                  public.employers e 
                  JOIN public.employers_industries ei ON e.id = ei.employer_id
                  JOIN public.industries i ON i.id = ei.industry_id
             GROUP BY i.name
             HAVING i.name = 'Разработка программного обеспечения'      
            '''
            
result = pd.read_sql_query(query_5_6, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1823736761.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_11, connection)


count
0   3553

Работодателей, занимающихся разработкой программного обесечения - 3553.

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [50]:
import requests
from bs4 import BeautifulSoup 

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров'
response = requests.get(url)

page = BeautifulSoup(response.text, 'html.parser')
mln_cities = tuple(a.text for a in page.find('table', class_="standard sortable").find_all('a')  if a.text[0].isupper())
mln_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [53]:
query_5_7 = f'''(SELECT 
                  a.name,
                  COUNT(v.id)
             FROM 
                  public.vacancies v
                  JOIN public.areas a ON v.area_id = a.id
                  JOIN public.employers e ON v.employer_id = e.id
             WHERE e.name = 'Яндекс' AND a.name IN {mln_cities}
             GROUP BY a.name)
             UNION ALL
             (SELECT 
                 'Total',
                  COUNT(v.id)
              FROM 
                  public.vacancies v
                  JOIN public.areas a ON v.area_id = a.id
                  JOIN public.employers e ON v.employer_id = e.id
              WHERE e.name = 'Яндекс' AND a.name IN {mln_cities})
              ORDER BY 2    
            '''
            
result = pd.read_sql_query(query_5_7, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\4015065792.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_12, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

Топ городов-миллионников России по числу вакансий "Яндекса" в них, по возрастанию: Омск (21 вакансия), Челябинск (22), Красноярск (23), Волгоград (24), Пермь (25), Казань (25), Ростов-на-Дону (25), Уфа (26), Самара (26), Краснодар (30), Воронеж (32), Новосибирск (35), Нижний Новгород (36), Екатеринбург (39), Санкт-Петербург (42), Москва (54). Всего в базе данных содержится информация о 485 вакансиях "Яндекса".

***

## Выводы по Юниту 5

Пятью крупнейшими работодателями в базе данных как по абсолютному количеству открытых вакансий, так и по количеству охватываемых вакансиями регионов являются (в порядке убывания) "Яндекс", "Ростелеком", "Тинькофф", "СБЕР" и "Газпром нефть". В первую очередь, это, конечно, даёт понять, какие отрасли экономики в настоящий момент являются самыми развитыми в России: IT-технологии ("Яндекс" и "Ростелеком"), банкинг ("Тинькофф" и "СБЕР") и нефтедобывающая промышленность ("Газпром нефть"). 

Кроме этого, полное совпадение верхушек рейтингов работодателей по количеству открываемых ими вакансий и по количеству регионов, охватывемых этими вакансиями, должно означать, что по меньшей мере крупные работодатели стремятся к равномерному распределению открываемых ими вакансий по регионам - то есть, чем больше новых работников нужно работодателям, тем в большем количестве регионов им эти новые работники требуются. Здравый смысл говорит,что это должно являться нормальной ситуацией - любой нормальный крупный бизнес должен стремиться к более широкой географической экспансии по мере своего собственного роста, а не концентрировать всю свою деятельность в своём родном и/или крупнейшем регионах страны.

У около 20% у работодателей в базе данных по какой-то причине отсутствует информация об их сфере деятельности. Это достаточно большое количество, чтобы привести к искажениям при использовании данной версии базы данных для какого-либо анализа сфер деятельности работодателей. Я рекомендую провести работу по заполнению пропусков в этом признаке, если вы хотите использовать предлагаемые данные для серьёзного исследования.

Всего лишь около четверти всех вакансий, открытых "Яндексом" (крупнейшим работодателем), были открыты в городах-миллионниках - т.е. абсолютное их большинство относятся к менее крупным городам. Скорее всего, это значит, что все рабочие места в филиалах компании, находящиеся в "миллионниках", уже заняты, а основная причина потребности в новых специалистах для работодателя в настоящее время - расширение компании в города со средней или маленькой численностью населения. Медианное (т.е. типичное) количество вакансий "Яндекса" в городе-миллионнике - 26. При сравнении "миллионников" друг с другом, в общем, количество вакансий "Яндекса" в них, грубо, но пропорционально численности их населения, что естественно.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [36]:
query_6_1 = '''SELECT COUNT(v.id)
             FROM public.vacancies v
             WHERE (LOWER(v.name) LIKE '%data%') OR (LOWER(v.name) LIKE '%данн%')      
            '''
            
result = pd.read_sql_query(query_6_1, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\3200626705.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_13, connection)


count
0   1771

1771 вакансия имеет отношение к данным.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [38]:
query_6_2 = '''SELECT COUNT(v.id)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'
                  AND ((LOWER(v.name) LIKE '%junior%') OR (v.experience = 'Нет опыта') OR (v.employment = 'Стажировка'))      
            '''
            
result = pd.read_sql_query(query_6_2, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1540436870.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_14, connection)


count
0     51

51 вакансия рассчитана на начинающего дата-саентиста.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [40]:
query_6_3 = '''SELECT COUNT(v.id)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'
                  AND (v.key_skills LIKE '%SQL%') OR(LOWER(v.key_skills) LIKE 'postgres')      
            '''
            
result = pd.read_sql_query(query_6_3, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\3243186619.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_15, connection)


count
0    201

Вакансий, имеющих отношение к SQL, в базе данных 201.

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [42]:
query_6_4 = '''SELECT COUNT(v.id)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'
                  AND LOWER(v.key_skills) LIKE '%python%'      
            '''
            
result = pd.read_sql_query(query_6_4, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\2742116260.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_16, connection)


count
0    351

Вакансий, имеющих отношение к Python, в базе данных 351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [73]:
query_6_5 = '''SELECT ROUND(AVG(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, '\t', '')) + 1), 2)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'      
            '''
            
result = pd.read_sql_query(query_6_5, connection)

result   

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\318583591.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_17, connection)


round
0   6.41

В среднем в вакансиях на дата-саентистов указывают шесть скиллов (точнее, 6.41 скилла).

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [77]:
query_6_6 = '''SELECT v.experience,
                     ROUND(AVG((COALESCE(v.salary_from, salary_to) + COALESCE(v.salary_to, salary_from)) / 2), 2)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'
                  AND ((v.salary_from IS NOT NULL) OR (v.salary_to IS NOT NULL)) 
            GROUP BY v.experience
            HAVING v.experience = 'От 3 до 6 лет'     
            '''
            
result = pd.read_sql_query(query_6_6, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1986582226.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_18, connection)


experience      round
0  От 3 до 6 лет  243114.67

Дата-саентист с опытом работы от трёх до шести лет может рассчитывть в среднем на заработную плату в 243 114.67 рублей.

Полезно будет узнать среднюю предлагаемую зарплату для всех категорий опыта:

In [79]:
query_6_7_bonus = '''SELECT v.experience,
                     ROUND(AVG((COALESCE(v.salary_from, salary_to) + COALESCE(v.salary_to, salary_from)) / 2), 2)
             FROM public.vacancies v
             WHERE 
                  (LOWER(v.name) LIKE '%data scientist%'
                  OR LOWER(v.name) LIKE '%машинн%обучен%'
                  OR v.name LIKE '%ML%'
                  OR LOWER(v.name) LIKE '%data science%'
                  OR LOWER(v.name) LIKE '%исследователь данных%'
                  OR LOWER(v.name) LIKE '%machine learning%')
                  AND v.name NOT LIKE '%HTML%'
                  AND ((v.salary_from IS NOT NULL) OR (v.salary_to IS NOT NULL)) 
            GROUP BY v.experience     
            '''
            
result = pd.read_sql_query(query_6_7_bonus, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\1089576882.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_18_bonus, connection)


experience      round
0           Нет опыта   74642.86
1  От 1 года до 3 лет  139674.75
2       От 3 до 6 лет  243114.67

***

## Выводы по Юниту 6

Вакансий, в сфере своей деятельности связанных с данными, в базе мало - всего около 2% от общего количества всех вакансий.
Поэтому ценность их для серьёзного анализа, скорее всего, невелика. Тем не менее, какие-то выводы по ним всё-таки сделать можно.

Мы видим, что ценность труда дата саентистов без опыта работы невелика - их средняя заработная плата составляет всего 75 тысяч рублей, что лишь немногим больше, чем средняя минимальная зарплата во всей базе данных (71 тысяча рублей). Такая низкая цена их труда, скорее всего, объясняется низким спросом на джуниоров: скорее всего, у большинства работодателей давно есть сотрудники, способные к элементарым операциям по обработке данных, и поэтому они предпочитают не нанимать лишних людей на эти должности (что подтверждается очень малой долей вакансий, открываемых на джуниоров - около 1/40 от всех вакансий на дата саентистов), а если приходится нанимать - предпочитают платить им как можно меньше, пока они не покажут себя в деле. Низкий спрос на джуниоров косвенно подтверждается малым количеством вакансий, рассчитанных на работу с Python или SQL (10% и 17% от всего объёма вакансий на дата саентистов, соответственно) - "базовыми" скиллами дата-саентистов, которыми часто и ограничавется скиллсет новичков, а также довольно большим средним количеством скиллов (шестью), требуемых работодателями в этой сфере от новых сотрудников.

Однако с приобретением ими даже минимального опыта, ценность труда дата саентистов резко возрастает. Уже для тех из них, кто обладает опытом работы от одного года до трёх лет, предлагаемая заработная плата в среднем составляет 140 тысяч рублей, что заметно больше среднего максимума предлагаемых по всей базе данных зарплат (110.5 тысяч рублей). А средняя заработная плата, предлагаемая дата саентистам, имеющим опыт работы от трёх до шести лет (243 тысячи рублей), превышает средний максимум для всех сфер деятельности более, чем в два раза(!). Это хорошо согласуется с тем, что работа с данными в настоящее время - всё ещё одна из самых перспективных и стабильно развивающихся сфер деятельности.

Интересно, что в базе данных полностью отстутствуют вакансии на дата саентистов с опытом работы больше шести лет. Скорее всего, найти такого опытного специалиста безработным практически нереально, и работодатели знают это.

# Общий вывод по проекту

## Самые значимые выводы, полученные в данном исследовании:

- На момент заполнения исследуемой базы данных цена на труд, скорее всего, была сильно повышена. В результате этого у более, чем половины вакансий,
  находящихся в базе, отсутствует информация о предлагаемой заработной плате - как результат того, что работодатели были вынуждены предлагать повышенную заработную плату нанимаемым сотрудникам на собеседовании, но не афишировать это публично, чтобы не демотивировать уже работающих у них сотрудников.

- Около трёх четвертей всех исследуемых вакансий предполагают традиционный тип занятости - полный график с полной занятостью. Ешё около одной пятой
  предполагают удалённую работу, почти все из этих вакансий предполагают полный график, но около одной девятой (или около 3% от общего количества вакансий) рассчитаны на работу по гибкому графику. Можно сделать вывод, что исследуемая база данных отражает растущую популярность удалённой работы. Доля остальных видов занятости в базе данных пренебрежительно мала.

- Около половины (53%) вакансий в базе данных рассчитаны на специалистов с опытом работы от одного до трёх лет. Намного меньше (около 32%) - на
  людей с опытом работы от трёх до шести лет. Очевидно желание большинства работодателей нанимать как можно больше "молодых" (с точки зрения стажа), но не совсем неопытных сотрудников. Скорее всего, это объясняется желанием работодателей платить меньшую зарплату - чем менее опытен работник, тем меньшей зарплаты он будет требовать. Однако при опыте работы меньше года, желание сэкономить на зарплате резко уступает недоверию неопытному сотруднику.

- Крупнейшим по количеству открытых вакансий работодателем в базе данных является "Яндекс". За ним идут (в порядке убывания) "Ростелеком", 
  "Тинькофф", "СБЕР" и "Газпром нефть". Эта пятёрка также является крупнейшей (в точно таком же порядке) по количеству регионов, в которых они размещают свои вакансии. Очевидно, крупнейшие работодатели равномерно распределяют свои вакансии по регионам, как следствие своего расширения через сети филиалов по всей территории страны.

- У около 20% у работодателей в базе данных по какой-то причине отсутствует информация об их сфере деятельности. Это достаточно большое количество,
  чтобы привести к искажениям при использовании данной версии базы данных для какого-либо анализа сфер деятельности работодателей. Я рекомендую провести работу по заполнению пропусков в этом признаке, если вы хотите использовать предлагаемые данные для серьёзного исследования.

- Всего лишь около четверти всех вакансий, открытых "Яндексом" (крупнейшим работодателем), были открыты в городах-миллионниках - т.е. абсолютное их 
  большинство относятся к менее крупным городам. Скорее всего, это значит, что все рабочие места в филиалах компании, находящиеся в "миллионниках", уже заняты, а основная причина потребности в новых специалистах для работодателя в настоящее время - расширение компании в города со средней или маленькой численностью населения.

- Труд специалистов по работе с данными в настоящее время всё ещё высоко ценится. Уже для тех из них, кто обладает опытом работы от одного года
  до трёх лет, предлагаемая заработная плата в среднем составляет 140 тысяч рублей, что заметно больше среднего максимума предлагаемых по всей базе данных зарплат (110.5 тысяч рублей). А средняя заработная плата, предлагаемая дата саентистам, имеющим опыт работы от трёх до шести лет (243 тысячи рублей), превышает средний максимум для всех сфер деятельности более, чем в два раза. 

# Дополнительное исследование

В качестве бонуса прикладываю исследование того, к каким ещё странам, помимо России, можно применить результаты вышеприведённого исследования. Наличие в списке регионов Алматы и Минска и здравый смысл позволяют предположить, что работодателям, пользующимся hh.ru, могут быть необходимы вакансии не только в России, но и в странах ближнего зарубежья. К ближнему зарубежью обычно относят республики бывшего СССР и иногда Финляндию (из-за близости к Санкт-Петербургу и наличия в этой стране русской диаспоры). Я буду считать представленными на сайте hh.ru в статистически значимом количестве те страны, в столицах которых были открыты вакансии в количестве не меньшем, чем один процент от всего объёма базы данных (что будет равно: 49 000 / 100 = 490).

In [57]:
closest_capitals = ('Киев', 'Минск', 'Алматы', 'Хельсинки', 'Бишкек', 'Душанбе', 'Ташкент',
                    'Ашхабад', 'Баку', 'Ереван', 'Тбилиси', 'Таллин', 'Рига', 'Вильнюс', 'Кишинёв')

query_bonus = f'''SELECT 
                  a.name,
                  COUNT(v.id)
             FROM 
                  public.vacancies v
                  JOIN public.areas a ON v.area_id = a.id
             WHERE a.name IN {closest_capitals}
             GROUP BY a.name
             HAVING COUNT(v.id) >= 490
             ORDER BY COUNT(v.id) DESC
             '''
             
result = pd.read_sql_query(query_bonus, connection)

result

C:\Users\Dmitriy\AppData\Local\Temp\ipykernel_2892\2358668274.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_1_bonus, connection)


name  count
0    Минск   2112
1   Алматы   1892
2  Ташкент    851

Итак, в БД в статистически значимом количестве содержится информация о вакансиях ещё в трёх странах, помимо России - в Белоруссии (Минск), 
Казахстане (Алматы) и Узбекистане (Ташкент). Это вполне естественно объясняется общим советским прошлым России с этими странами, всё ещё дающим о себе знать в экономических связях, уровне взаимной иммиграции, и распространённости русского языка в них. Для данного исследования же это будет значить, что его результаты, правда скорее всего - с некоторыми поправками, возможно будут применимы не только для России, но также и для этих трёх стран. 